### PR TITLE
[CDAP-21079] Implement cdap-messaging-spi extension for spanner - Part 1

### PR DESCRIPTION
**Pull request by sidhdirenge**
Implement methods like createTopic(), publish() etc.
Follow up PR to include changes for fetch().

The code is tested in autopush project.

image

Junits are tested by setting the system properties, as done in SpannerStructuredAdminTest.
Screenshot 2024-12-09 at 2 04 58 PM
image

### CODE REVIEW

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/client/DelegatingMessagingService.java [hunk 1]
```diff
@@ -139,13 +142,25 @@ private MessagingService getDelegate() {
       LOG.info("Messaging service {} is loaded", messagingService.getName());
       try {
         messagingService.initialize(new DefaultMessagingServiceContext(this.cConf));
-      } catch (IOException e) {
+        createTopics(cConf, messagingService);
```
Comment (masoud-io) [inline cdap-tms/src/main/java/io/cdap/cdap/messaging/client/DelegatingMessagingService.java]: I assume we need this for Spanner implementation. It's a bit strange to create topic here. How are we creating topics for TMS implementation now?

Comment (masoud-io) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/client/DelegatingMessagingService.java]: Why not putting this in initialize() method inside spanner implementation?

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/service/CoreMessagingService.java
Comment (masoud-io) [inline cdap-tms/src/main/java/io/cdap/cdap/messaging/service/CoreMessagingService.java:1]: Revert these changes. Nothing has been changed in this file.

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+
+  private DatabaseAdminClient adminClient;
+
+  private String instanceId;
+
+  private String databaseId;
+
+  private final ConcurrentLinkedQueue<StoreRequest> batch = new ConcurrentLinkedQueue<>();
+
+  public static final String PAYLOAD_FIELD = "payload";
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: move static finals above field definitions.

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    return String.format("CREATE TABLE IF NOT EXISTS %s ( %s INT64, %s INT64, %s"
+            + " TIMESTAMP NOT NULL OPTIONS (allow_commit_timestamp=true), %s BYTES(MAX) )"
+            + " PRIMARY KEY (%s, %s, %s), ROW DELETION POLICY"
+            + " (OLDER_THAN(publish_ts, INTERVAL 7 DAY))", getTableName(topicId), SEQUENCE_ID_FIELD,
+        PAYLOAD_SEQUENCE_ID, PUBLISH_TS_FIELD, PAYLOAD_FIELD, SEQUENCE_ID_FIELD,
+        PAYLOAD_SEQUENCE_ID, PUBLISH_TS_FIELD);
+  }
+
+  public static String getTableName(TopicId topicId) {
+    return topicId.getNamespace() + topicId.getTopic();
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: In addition to namespace, I recommend to prefix these tables such that they can easily be distinguished with other tables. Example messaging_namespace_topicId

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        .to(topicId).set(PROPERTIES_FIELD).to(Value.json(jsonString)).build();
+
+    try {
+      client.write(Collections.singleton(mutation));
+    } catch (SpannerException e) {
+      if (e.getErrorCode() == ErrorCode.NOT_FOUND) {
+        throw new TopicNotFoundException(topicMetadata.getTopicId().getNamespace(),
+            topicMetadata.getTopicId().getTopic());
+      }
+      LOG.error("Failed to update topic {}", topicId, e);
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: It's not a good practice to both log and throw an exception. I'd only throw an exception here and let client handles the exception.

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+        StoreRequest headRequest = batch.poll();
+        for (byte[] payload : headRequest) {
+          Mutation mutation = Mutation.newInsertBuilder(getTableName(headRequest.getTopicId()))
+              .set(SEQUENCE_ID_FIELD).to(i++).set(PAYLOAD_SEQUENCE_ID).to(0).set(PUBLISH_TS_FIELD)
+              .to("spanner.commit_timestamp()").set(PAYLOAD_FIELD).to(ByteArray.copyFrom(payload))
+              .build();
+          batchCopy.add(mutation);
+        }
+
+        if (batch.isEmpty() && (i < 50 || System.currentTimeMillis() - start < 50)) {
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: Set 50 in config

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+          try {
+            Thread.sleep(5);
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: set 5 in config

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+            LOG.error("error during sleep", e);
+            throw new IOException(e);
+          }
+        }
+      }
+      if (!batchCopy.isEmpty()) {
+        try {
+          client.write(batchCopy);
+        } catch (SpannerException e) {
+          LOG.error("failed to publish message ", e);
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: Same here. Do not log and throw

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/client/DelegatingMessagingService.java (hunk 1 above)
Comment (sidhdirenge) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/client/DelegatingMessagingService.java]: For TMS implementation, it is part of the start up which is called when MessagingService pod comes up : [CoreMessagingService.java#L239](https://github.com/cdapio/cdap/blob/develop/cdap-tms/src/main/java/io/cdap/cdap/messaging/service/CoreMessagingService.java#L239)

I was planning to do createTopics() in SpannerMessagingService.initialise() but the issue here is the MessagingServiceUtils class is also in cdap-tms. Also, DefaultTopicMetadata which is input to the createTopic() method. This class is in cdap-tms module (same level as the DefaultMessagingServiceContext, DelegatingMessagingService). 
We can somehow get away with the DefaultTopicMetadata dependency, as the properties which are set here are not very useful in case of Cloud Spanner. But the MessagingServiceUtils is needed here.

I have added some code comments in the createTopics(), may be, I can add a TODO as well there stating this method can be removed from DelegatingMessagingService after the cdap-tms module refactoring.

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java (hunk 6 above)
Comment (sidhdirenge) [reply cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: Done.

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/client/DelegatingMessagingService.java (hunk 1 above)
Comment (sidhdirenge) [reply cdap-tms/src/main/java/io/cdap/cdap/messaging/client/DelegatingMessagingService.java]: I have updated code a bit, to just extract TopicMetadata list in Delegating service and then to createTopics() in Spanner implementation.
PTAL.

File: cdap-common/src/main/resources/cdap-default.xml [hunk 8]
```diff
@@ -247,7 +247,10 @@
 
   <property>
     <name>twill.jvm.gc.opts</name>
-    <value>-XX:+UseG1GC -verbose:gc -Xloggc:&lt;LOG_DIR&gt;/gc.log -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+UseGCLogFileRotation -XX:NumberOfGCLogFiles=10 -XX:GCLogFileSize=1M</value>
+    <value>-XX:+UseG1GC -verbose:gc -Xloggc:&lt;LOG_DIR&gt;/gc.log -XX:+PrintGCDetails
```
Comment (masoud-io) [inline cdap-common/src/main/resources/cdap-default.xml]: Is this formatting change necessary?

File: cdap-common/src/main/resources/cdap-default.xml [hunk 9]
```diff
@@ -2661,6 +2664,31 @@
...
+    <description>
+      The default maximum waiting time before a batch is published by spanner messaging service.
+    </description>
+  </property>
+
+  <property>
+    <name>messaging.spanner.properties.publish.delay.millis</name>
+    <value>5</value>
+    <description>
+      The default waiting time for receiving new publish calls that ensures efficient batching and
```
Comment (masoud-io) [inline cdap-common/src/main/resources/cdap-default.xml]: Update the description. We cannot control publish calls. Clients call publish.

File: cdap-messaging-spi/src/main/java/io/cdap/cdap/messaging/spi/MessagingService.java [hunk 10]
```diff
@@ -39,9 +39,10 @@ public interface MessagingService {
    * Initialize the messaging service. This method is guaranteed to be called before any other
    * method is called. It will only be called once for the lifetime of the messaging service.
    *
-   * @param context the context that can be used to initialize the messaging service.
+   * @param context      the context that can be used to initialize the messaging service.
+   * @param systemTopics topics to be created initially.
    */
-  void initialize(MessagingServiceContext context) throws IOException;
+  void initialize(MessagingServiceContext context, List<TopicMetadata> systemTopics) throws IOException;
```
Comment (masoud-io) [inline cdap-messaging-spi/src/main/java/io/cdap/cdap/messaging/spi/MessagingService.java]: Better to pass systemTopics through MessagingServiceContext as context already has cConf and AFAICT topicsToCreate() just needs cConf. So we just need to add `getSystemTopics` method in context. 
WDYT?

Comment (sidhdirenge) [reply cdap-messaging-spi/src/main/java/io/cdap/cdap/messaging/spi/MessagingService.java]: Thats a nice suggestion. Updated the code accordingly.
Thanks!

File: cdap-common/src/main/resources/cdap-default.xml [hunk 11]
```diff
@@ -2661,6 +2661,31 @@
...
+  <property>
+    <name>messaging.spanner.properties.publish.batch.timeout.millis</name>
+    <value>50</value>
+    <description>
+      The default maximum waiting time before a batch is published by spanner messaging service.
+    </description>
+  </property>
+
+  <property>
+    <name>messaging.spanner.properties.publish.delay.millis</name>
```
Comment (masoud-io) [inline cdap-common/src/main/resources/cdap-default.xml]: I would not call this delay. AFAICT, this is polling frequency while either batch.size is met or batch.timeout is met. How about `messaging.spanner.properties.publish.batch.poll.millis` ?

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+    for (TopicMetadata topic : topics) {
+      LOG.info("Creating topic : {}", topic.getTopicId().getTopic());
+      ddlStatements.add(getCreateTopicDDLStatement(topic.getTopicId()));
+    }
+    executeCreateDDLStatements(ddlStatements);
+    updateTopicMetadataTable(topics);
+    LOG.info("Created all system topics.");
+  }
+
+  private static String getCreateTopicMetadataDDLStatement() {
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: nit: no need to make it static.

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+      }
+      if (!batchCopy.isEmpty()) {
+        try {
+          client.write(batchCopy);
+        } catch (SpannerException e) {
+          throw new IOException(e);
+        }
+      }
+    }
+    return null;
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java:324]: Better to return a RollbackDetail implementation that throws exceptions if any of its methods is called.

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+    }
+
+    @Override
+    public void close() {
+      resultSet.close();
+    }
+  }
+
+  public static byte[] getMessageId(long sequenceId, long messageSequenceId, long timestamp) {
+    LOG.info("sequenceId {} messageSequenceId {} timestamp {}", sequenceId, messageSequenceId,
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: This is gonna be super chatty. If you think it's critical in the future for debugging purposes, set it to trace.

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 15]
```diff
@@ -0,0 +1,410 @@
...
+        timestamp);
+    byte[] result = new byte[Bytes.SIZEOF_LONG + Bytes.SIZEOF_SHORT + Bytes.SIZEOF_LONG
+        + Bytes.SIZEOF_SHORT];
+    int offset = 0;
+    // Implementation copied from MessageId
+    offset = Bytes.putLong(result, offset, timestamp);
+    offset = Bytes.putShort(result, offset, (short) sequenceId);
+    offset = Bytes.putLong(result, offset, 0);
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: Add comments why 0 here.

File: cdap-tms/src/main/java/io/cdap/cdap/messaging/client/DefaultMessagingServiceContext.java [hunk 16]
```diff
@@ -38,11 +45,23 @@ public class DefaultMessagingServiceContext implements MessagingServiceContext {
   @Override
   public Map<String, String> getProperties() {
     // TODO: cdap-tms module refactoring will remove this dependency on spanner.
```
Comment (masoud-io) [inline cdap-tms/src/main/java/io/cdap/cdap/messaging/client/DefaultMessagingServiceContext.java]: Add JIRA ticket

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java (hunk 13 above)
Comment (sidhdirenge) [reply cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java:324]: Added TODO since if we throw exception here, method signature needs to be updated.

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java (hunk 14 above)
Comment (sidhdirenge) [reply cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: Not part of this PR, reverted the change

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java (hunk 15 above)
Comment (sidhdirenge) [reply cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: Not part of this PR, reverted the change

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+
+  @Override
+  public void createTopic(TopicMetadata topicMetadata)
+      throws TopicAlreadyExistsException, IOException, UnauthorizedException {
+    List<String> ddlStatements = new ArrayList<>();
+    ddlStatements.add(getCreateTopicMetadataDDLStatement());
+    ddlStatements.add(getCreateTopicDDLStatement(topicMetadata.getTopicId()));
+    executeCreateDDLStatements(ddlStatements);
+    updateTopicMetadataTable(Collections.singletonList(topicMetadata));
+    LOG.info("Created topic : {}", topicMetadata.getTopicId().getTopic());
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: Why not trace?

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+   */
+  @Nullable
+  @Override
+  public RollbackDetail publish(StoreRequest request)
+      throws TopicNotFoundException, IOException, UnauthorizedException {
+    long start = System.currentTimeMillis();
+
+    batch.add(request);
+    if (!batch.isEmpty()) {
+      int i = 0;
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: rename to sequenceId

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+      List<Mutation> batchCopy = new ArrayList<>(batch.size());
+      // We need to batch less than fetch limit since we read for publish_ts >= last_message.publish_ts
+      // see fetch for explanation of why we read for publish_ts >= last_message.publish_ts and
+      // not publish_ts > last_message.publish_ts
+      while (!batch.isEmpty()) {
+        StoreRequest headRequest = batch.poll();
+        for (byte[] payload : headRequest) {
+          Mutation mutation = Mutation.newInsertBuilder(getTableName(headRequest.getTopicId()))
+              .set(SEQUENCE_ID_FIELD).to(i++).set(PAYLOAD_SEQUENCE_ID).to(0).set(PUBLISH_TS_FIELD)
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: PAYLOAD_SEQUENCE_ID is always zero. This only work for payloads up to certain size. 
The limit is 10 MiB per cell in spanner: https://cloud.google.com/spanner/quotas#tables
We do need to break up payloads and insert them as two rows with PAYLOAD_SEQUENCE_ID=0 and PAYLOAD_SEQUENCE_ID=1, PAYLOAD_SEQUENCE_ID=2, etc.

This can be done in a separate PR. Open a Jira and put a link here.

Comment (masoud-io) [reply cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: Also there's also a limit request size. You should make sure that this batching mechanism does not lead to batch size growing beyond the limit. If so, what will happen is that request will be rejected and txn will never commit. 
So, you need to keep calculating the size of messages being inserted into batchCopy and once you reach a limit, commit regardless of batch logic (e.g., polling and waiting for certain period of time or certain number of messages).

Comment (sidhdirenge) [reply cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: Opened JIRA : https://cdap.atlassian.net/browse/CDAP-21094

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 20]
```diff
@@ -0,0 +1,367 @@
...
+
+  @Override
+  public String getName() {
+    return this.getClass().getSimpleName();
+  }
+
+  @Override
+  public void createTopic(TopicMetadata topicMetadata)
+      throws TopicAlreadyExistsException, IOException, UnauthorizedException {
+    List<String> ddlStatements = new ArrayList<>();
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: You don't need additional implementation here. Simply call createTopic(ImmutableList.of(topicMetadata)).

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 21]
```diff
@@ -0,0 +1,367 @@
...
+    }
+  }
+
+  @Override
+  public void deleteTopic(TopicId topicId)
+      throws TopicNotFoundException, IOException, UnauthorizedException {
+
+    String topicTableName = getTableName(topicId);
+    String deleteTopicTableSQL = String.format("DROP TABLE IF EXISTS %s", topicTableName);
+    OperationFuture<Void, UpdateDatabaseDdlMetadata> future = adminClient.updateDatabaseDdl(
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: move this into try

File: cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java [hunk 22]
```diff
@@ -0,0 +1,367 @@
...
+    }
+    try {
+      client.write(mutations);
+    } catch (SpannerException e) {
+      throw new IOException(e);
+    }
+  }
+
+  private void executeCreateDDLStatements(List<String> ddlStatements) throws IOException {
+    OperationFuture<Void, UpdateDatabaseDdlMetadata> future = adminClient.updateDatabaseDdl(
```
Comment (masoud-io) [inline cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java]: move this into try

**Review by masoud-io — APPROVED** [review]
Couples of more comments. Otherwise LGTM
